workflows/sampletracking.nf

```diff
@@ -40,12 +43,78 @@ workflow SAMPLETRACKING {
     def ch_multiqc_files = Channel.empty()
     def ch_pool_multiqc_files = Channel.empty()
 
+
+    def (ch_sample, ch_snp, ch_rest) = ch_samplesheet
+        .multiMap {meta, sample_bam, sample_bam_index, snp_fastq, snp_bam, snp_bam_index ->
+            sample: [meta, (sample_bam instanceof List && sample_bam.isEmpty() ? null : sample_bam), (sample_bam_index ?: null)]
+            snp   : [meta, (snp_bam   instanceof List && snp_bam.isEmpty()   ? null : snp_bam),     (snp_bam_index   ?: null)]
+            rest  : [meta, (snp_fastq instanceof List && snp_fastq.isEmpty() ? null : snp_fastq)]
+        }
+
+    def (ch_sample_with_idx, ch_sample_no_idx) = ch_sample.branch {
+        _meta, _sample_bam, sample_bam_index ->
+            with_index: sample_bam_index != null
+            no_index  : sample_bam_index == null
+    }
+
+    ch_sample_no_idx = ch_sample_no_idx.filter { _meta, bam, _idx -> bam != null }
+
+    SAMTOOLS_INDEX(
+        ch_sample_no_idx.map { meta, sample_bam, _sample_bam_index -> [meta, sample_bam] }
+    )
+
+    def ch_sample_idx_all = SAMTOOLS_INDEX.out.bai
+        .mix(SAMTOOLS_INDEX.out.csi)
+        .mix(SAMTOOLS_INDEX.out.crai)
+
+    def ch_sample_fixed = ch_sample_with_idx.mix(
+        ch_sample_no_idx
+            .join(ch_sample_idx_all, by: 0)
+            .map { meta, sample_bam, _old_index, new_index -> [meta, sample_bam, new_index] }
+    )
+
+    def (ch_snp_with_idx, ch_snp_no_idx, ch_snp_none) = ch_snp.branch {
+        _meta, snp_bam, snp_bam_index ->
+            with_index: snp_bam_index != null && snp_bam != null
+            no_index  : snp_bam_index == null && snp_bam != null
+            none      : snp_bam == null
+    }
+
+    ch_snp_no_idx    = ch_snp_no_idx.filter    { _meta, bam, _idx -> bam != null }
+
+    SAMTOOLS_INDEX_SNP_BAM(
+        ch_snp_no_idx.map { meta, snp_bam, _snp_bam_index -> [meta, snp_bam] }
+    )
+
+    def ch_snp_idx_all = SAMTOOLS_INDEX_SNP_BAM.out.bai
+        .mix(SAMTOOLS_INDEX_SNP_BAM.out.csi)
+        .mix(SAMTOOLS_INDEX_SNP_BAM.out.crai)
+
+    def ch_snp_fixed = ch_snp_with_idx.mix(
+        ch_snp_no_idx
+            .join(ch_snp_idx_all, by: 0)
+            .map { meta, snp_bam, _old_index, new_index -> [meta, snp_bam, new_index] }
+    ).mix(
+        ch_snp_none.map { meta, _bam, _idx -> [meta, null, null] }
```

Comment:
```suggestion
        ch_snp_none
```
The default `[]` values should be retained to prevent failures down the line in the pipeline